cyclos_configuration/scripts/direct_debits/Library.groovy

```diff
@@ -0,0 +1,457 @@
+		batchFields.batchId = pain_008.batchId
+		batchFields.xml = xml
+        batchFields.totalAmount = pain_008.totalAmount
+        batchFields.nrOfTrxs = pain_008.trxs.size()
+		recordService.saveEntity(data.dto)
+
+        // Update the status of the records that are succesfully added to the PAIN.008 string: change status open to submitted and retry to resubmitted.
+        // Also store the batchId in the changed records.
+        records.each { RecordVO recordVO ->
+			def recordDTO = recordService.load(recordVO.id)
+            def record = scriptHelper.wrap(recordDTO)
+            record.batchId = pain_008.batchId
+            record.status = (record.status as CustomFieldPossibleValue).internalName == 'open' ? 'submitted' : 'resubmitted'
+            recordService.save(recordDTO)
+        }
+
+        return "${records.size()} direct debits were processed succesfully in batch ${pain_008.batchId}."
+    }
+
+    /**
+     * Returns a list of directDebit records with status open or retry. These are the records that should be 
+     * included in the next PAIN.008 batch.
+     */
+    private List<RecordVO> _getDirectDebitRecords() {
+		def query = new UserRecordQuery()
+		query.type = new RecordTypeVO(internalName: 'directDebit')
+		query.customValues = [
+			new CustomFieldValueForSearchDTO(
+				field: new CustomFieldVO(internalName: 'directDebit.status'),
+				enumeratedValues: [
+					new CustomFieldPossibleValueVO(internalName: 'open'),
+					new CustomFieldPossibleValueVO(internalName: 'retry')
+				] as Set
+			)
+		] as Set
+		query.setSkipTotalCount(true)
+		query.setUnlimited()
+        return recordService.search(query).pageItems
+    }
+
+	/**
+	 * Returns the newest valid eMandate record for the given user.
+     * Note: we can not use the current() method in the eMandate library, because this returns the newest record, even if this is not active yet.
+     * Here we need the newest active eMandate.
+	 */
+	private SystemRecord _getEMandate(User user) {
+		def query = new SystemRecordQuery()
+		query.type = new RecordTypeVO(internalName: 'eMandate')
+		query.customValues = [
+			new CustomFieldValueForSearchDTO(
+				field: new CustomFieldVO(internalName: 'eMandate.owner'),
+				linkedEntityValues: [
+					new LinkedEntityVO(user.id)
+				] as Set
+			),
+			new CustomFieldValueForSearchDTO(
+				field: new CustomFieldVO(internalName: 'eMandate.status'),
+				enumeratedValues: [
+					new CustomFieldPossibleValueVO(internalName: 'success')
+				] as Set
+			)
+		] as Set
+		query.setPageSize(1) // Since the result is by default ordered by creation date, this gives us the newest record.
+		def results = recordService.search(query).pageItems
+		return results.isEmpty() ? null : entityManagerHandler.find(SystemRecord, results[0].id)
+	}
+	
+    /**
+     * Returns a list of all direct debit user records of the given user.
+     */
+    List<UserRecord> getDirectDebits(User user) {
+        def query = new UserRecordQuery()
```

Comment:
Is it also possible to let this query skip records with a certain custom field value? I know I can add a construction using CustomFieldValueForSearchDTO to search for certain custom field values, just like in the _getEMandate() method, right above, but I don't know how to search for _anything but_ certain custom field values. It is not very important, I just wondered if there is an easy way to do this.

Reply:
There's no way to exclude users by custom fields. You'd need to do this manually.